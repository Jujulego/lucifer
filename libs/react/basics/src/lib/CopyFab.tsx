import React, { MouseEvent, useState, useEffect, FC } from 'react';
import copy from 'copy-to-clipboard';
import clsx from 'clsx';

import { Fab, FabProps, FabClassKey, Fade } from '@material-ui/core';
import { makeStyles, Theme } from '@material-ui/core/styles';
import CheckIcon from '@material-ui/icons/Check';
import FileCopyIcon from '@material-ui/icons/FileCopy';

import { StyledProps } from '@lucifer/react-utils';

// Types
export type CopyFabClassKey = FabClassKey | 'succeed';
export interface CopyFabProps extends Omit<FabProps, 'classes' | 'children'>, StyledProps<CopyFabClassKey> {
  /** Content to be copied */
  text: string;

  /**
   * Mime type of content
   * @default text/plain
   */
  format?: string;

  /** Copied callback, called after a successful copy */
  onCopied?: () => void;
}

// Styles
const useStyles = makeStyles(({ palette }: Theme) => ({
  label: {
    display: 'grid',
    justifyItems: 'center',

    '& > *': {
      gridArea: '1 / 1 / 2 / 2',
    }
  },
  succeed: {
    backgroundColor: palette.success.main,

    '&:hover': {
      backgroundColor: palette.success.dark,
    }
  }
}));

// Component
const CopyFab: FC<CopyFabProps> = (props) => {
  // Props
  const {
    text, classes,
    format = "text/plain",
    onCopied, onClick,
    ...btn
  } = props;

  // State
  const [succeed, setSucceed] = useState(false);

  // Effects
  useEffect(() => {
    if (succeed) {
      const id = setTimeout(() => setSucceed(false), 2000);
      return () => clearTimeout(id);
    }
  }, [succeed, setSucceed]);

  // Handler
  const handleCopy = (event: MouseEvent<HTMLButtonElement>) => {
    if (onClick) onClick(event);

    // Copy
    const success = copy(text, { format });
    if (success) {
      setSucceed(true);
      if (onCopied) onCopied();
    }
  };

  // Render
  const styles = useStyles({
    classes: {
      succeed: classes?.succeed,
      label: classes?.label
    }
  });

  return (
    <Fab
      {...btn} onClick={handleCopy}
      classes={{
        ...classes,
        root: clsx(classes?.root, { [styles.succeed]: succeed }),
        label: styles.label,
      }}
    >
      <Fade in={succeed} timeout={250}>
        <CheckIcon />
      </Fade>
      <Fade in={!succeed} timeout={250}>
        <FileCopyIcon fontSize="small" />
      </Fade>
    </Fab>
  );
};

export default CopyFab;
