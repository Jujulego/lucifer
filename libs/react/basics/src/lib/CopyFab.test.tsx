import React from 'react';
import copy from 'copy-to-clipboard';
import { screen, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import CopyFab from '../lib/CopyFab';

// Mocks
jest.mock('copy-to-clipboard');

beforeAll(() => {
  jest.useFakeTimers();
});

afterAll(() => {
  jest.restoreAllMocks();
});

beforeEach(() => {
  (copy as jest.Mock).mockReset();
});

// Tests
describe('CopyFab', () => {
  it('should render a button', () => {
    // Render
    render(
      <CopyFab text="test" />
    );

    // Check elements
    expect(screen.queryByRole('button')).toBeInTheDocument();
  });

  it('should copy on click', () => {
    const spy = jest.fn();
    (copy as jest.Mock).mockImplementation(() => true);

    // Render
    render(
      <CopyFab text="test" onCopied={spy} />
    );

    // Interact
    const btn = screen.getByRole('button');
    userEvent.click(btn);

    // Check callbacks
    expect(spy).toBeCalled();
    expect(copy).toHaveBeenCalledWith('test', { format: 'text/plain' });
  });
});
