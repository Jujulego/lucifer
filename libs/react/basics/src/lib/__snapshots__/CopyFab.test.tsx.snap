// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<WithStyles(ForwardRef(Fab))
  classes={
    Object {
      "label": "makeStyles-label-1",
      "root": "",
    }
  }
  onClick={[Function]}
>
  <ForwardRef(Fade)
    in={false}
    timeout={250}
  >
    <Memo([object Object]) />
  </ForwardRef(Fade)>
  <ForwardRef(Fade)
    in={true}
    timeout={250}
  >
    <Memo([object Object])
      fontSize="small"
    />
  </ForwardRef(Fade)>
</WithStyles(ForwardRef(Fab))>
`;
